internal/graph/check.go

```diff
@@ -522,6 +533,33 @@ func (c *LocalChecker) checkDirect(parentctx context.Context, req *ResolveCheckR
 			err = validation.ValidateTuple(typesys, tk)
 
 			if t != nil && err == nil {
+				typeRestrictions, err := typesys.GetDirectlyRelatedUserTypes(objectType, relation)
+				if err != nil {
+					return nil, err
+				}
+
+				var condition *condition.EvaluableCondition
+				for _, typeRestriction := range typeRestrictions {
```

Comment:
We will need to evaluate a condition anywhere we read a tuple from the storage layer, not just on a direct relationship evaluations. Here are a few good examples,

```
type user

type group
  relations
    define member: [user]

type document
  relations
    define viewer: [group#member with condition1]

condition condition1(x int)
   x < 100

- document:1#viewer@group:eng#member
- group:eng#member@user:jon

Check(document:1#viewer@user:jon) should return {allowed: false} because the group membership relationship is conditioned on x < 100, and that condition is not met so the userset (not direct) relationship should be discredited. 
```

Reply:
In https://github.com/openfga/openfga/pull/999/commits/088389df22eda7819e7d8634568201fd9bcfe8ca I moved validation logic used in Write into ValidateTuple which would be reused wherever we're filtering out invalid tuples. Haven't yet tried those changes against these.